lib/commands/audit.js

```diff
@@ -1,8 +1,389 @@
+    }
+
+    await Promise.all(edges.map((edge) => this.getVerifiedInfo(edge)))
+
+    // Sort alphabetically
+    const invalid = Array.from(this.invalid).sort((a, b) => localeCompare(a.name, b.name))
+    const missing = Array.from(this.missing).sort((a, b) => localeCompare(a.name, b.name))
+
+    const verified = invalid.length === 0 && missing.length === 0
+
+    if (!verified) {
+      this.exitCode = 1
+    }
+
+    const end = process.hrtime.bigint()
+    const elapsed = end - start
+
+    if (this.npm.config.get('json')) {
+      this.appendOutput(this.makeJSON({ invalid, missing }))
+    } else {
+      const timing = `audited ${this.audited} packages in ${Math.floor(Number(elapsed) / 1e9)}s`
+      const verifiedPrefix = verified ? 'verified registry signatures, ' : ''
+      this.appendOutput(`${verifiedPrefix}${timing}\n`)
+
+      if (this.verified && !verified) {
+        this.appendOutput(
+          `${this.verified} packages have ${this.npm.color ? chalk.bold('verified') : 'verified'}` +
+          ` registry signatures\n`
+        )
+      }
+
+      if (missing.length) {
+        const msg = missing.length === 1 ?
+          `package has a ${this.npm.color ? chalk.bold(chalk.magenta('missing')) : 'missing'}` +
+          ` registry signature` :
+          `packages have ${this.npm.color ? chalk.bold(chalk.magenta('missing')) : 'missing'}` +
+          ` registry signatures`
+        this.appendOutput(
+          `${missing.length} ${msg} but the registry is ` +
+          `providing signing keys${this.npm.config.get('missing') ? ':\n' : ''}`
+        )
+        // TODO: This might not be the right option for this
+        if (this.npm.config.get('missing')) {
+          this.appendOutput(this.humanOutput(missing))
+        } else {
+          this.appendOutput(`  run \`npm audit signatures --missing\` for details`)
+        }
+      }
+
+      if (invalid.length) {
+        const msg = invalid.length === 1 ?
+          `package has an ${this.npm.color ? chalk.bold(chalk.red('invalid')) : 'invalid'}` +
+          ` registry signature` :
+          `packages have ${this.npm.color ? chalk.bold(chalk.red('invalid')) : 'invalid'}` +
+          ` registry signatures`
+        this.appendOutput(
+          `${missing.length ? '\n' : ''}${invalid.length} ${msg}:\n`
+        )
+        this.appendOutput(this.humanOutput(invalid))
+        const plural = invalid.length === 1 ? '' : 's'
+        this.appendOutput(
+          `\nSomeone might have tampered with the package${plural} ` +
+          `since it was published on the registry (monster-in-the-middle attack)!\n`
+        )
+      }
+    }
+  }
+
+  findAllRegistryUrls (edges, opts) {
+    return Array.from(edges.reduce((p, edge) => {
+      let alias = false
+      try {
+        alias = npa(edge.spec).subSpec
+      } catch (err) {
+      }
+      const spec = npa(alias ? alias.name : edge.name)
+      p.add(fetch.pickRegistry(spec, opts))
+      return p
+    }, new Set()))
```

Comment:
```suggestion
    return new Set(Array.from(edges, (edge) => {
      let alias = false
      try {
        alias = npa(edge.spec).subSpec
      } catch (err) {
      }
      const spec = npa(alias ? alias.name : edge.name)
      return fetch.pickRegistry(spec, opts)
    }))
```
this will work whether edges is an array, or a Set (see above comment)